### DISCUSSION TITLE
Worker Address Port for TCPIP Connections

### DISCUSSION THREAD

**Discussion by mstaylor** [openucx/ucx#9258] [Q&A]
We are trying to run UCX on AWS via ECS Docker containers on Cylon (https://github.com/cylondata/cylon).  Our configuration utilizes the UCP address created by the ucp_worker_get_address command.  Is there a capability to prevent automatic port assignment and thus provide more declarative control over the address generation process?

**Comment by changchengx** [comment]
How to understand ```prevent automatic port assignment```? Anymore information?

**Reply by mstaylor** [reply]
The underlying issue is we need to map ip and port address from a given container to a host in docker systems (ECS in this case) and this port is dynamically assigned by UCP. In otherwords, there is no mechanism during runtime execution that we can find to derive that address when our code calls ucp_worker_get_address and we subsequently pass that worker address to ucp_ep_params_t.  So our code looks like this:

epParams.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS |
                              UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
        epParams.address = address;
        epParams.err_mode = UCP_ERR_HANDLING_MODE_NONE;

        // Create an endpoint
        ucxStatus = ucp_ep_create(comm.ucpSendWorker, &epParams, &ep);


The above address is the worker address retrieved from the ucp_worker_get_address function.

The problem is we need the capability to specify the port number when the UCP address is created for the worker.  Is that possible?  I've gone through examples and the source code itself and the only thing I can find is to disregard the worker address and use the socket addr handler (from the https://github.com/openucx/ucx/tree/master/examples/ucp_client_server.c code):

static ucs_status_t
start_server(ucp_worker_h ucp_worker, ucx_server_ctx_t *context,
             ucp_listener_h *listener_p, const char *address_str)
{
    struct sockaddr_storage listen_addr;
    ucp_listener_params_t params;
    ucp_listener_attr_t attr;
    ucs_status_t status;
    char ip_str[IP_STRING_LEN];
    char port_str[PORT_STRING_LEN];

    set_sock_addr(address_str, &listen_addr);

    params.field_mask         = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR |
                                UCP_LISTENER_PARAM_FIELD_CONN_HANDLER;
    params.sockaddr.addr      = (const struct sockaddr*)&listen_addr;
    params.sockaddr.addrlen   = sizeof(listen_addr);

**Comment by changchengx** [comment]
> The problem is we need the capability to specify the port number when the UCP address is created for the worker. Is that possible?
>
As far as I know, it doesn't exist.
To exchange the ucp_address_t, it could use internal connection manager or out-off-bound transport(e.g. tcp server/client). If the app could only know the ```port``` number after getting the ucp_address, there's no functionality to assign the port.
